lib/pbio/drv/gpio/gpio_ev3.c

```diff
@@ -31,13 +34,30 @@ static void pbdrv_gpio_alt_gpio(const pbdrv_gpio_t *gpio) {
     pbdrv_gpio_alt(gpio, mux_info->gpio_mode);
 }
 
+// If the pin is not in banks 0/1, the PRU is not involved.
+// Otherwise, if the PRU is initialized, do direction changes
+// via the PRU in order to prevent race conditions.
+// If the PRU is not initialized (i.e. doing direction changes
+// during early boot), also set the direction directly via
+// the ARM.
+#define PBDRV_GPIO_EV3_ARM_OWNS_GPIO_BANK(pin_index)    \
+    ((pin_index) > 32 || !pbdrv_rproc_is_ready())
+
 static void gpio_write(const pbdrv_gpio_t *gpio, uint8_t value) {
     if (!gpio) {
         return;
     }
     pbdrv_gpio_alt_gpio(gpio);
     uint32_t pin_index = get_pin_index(gpio);
-    GPIODirModeSet(SOC_GPIO_0_REGS, pin_index, GPIO_DIR_OUTPUT);
+    if (PBDRV_GPIO_EV3_ARM_OWNS_GPIO_BANK(pin_index)) {
+        GPIODirModeSet(SOC_GPIO_0_REGS, pin_index, GPIO_DIR_OUTPUT);
+    } else {
+        uint32_t val = 1 << (pin_index - 1);
+        pbdrv_rproc_ev3_pru1_shared_ram.gpio_bank_01_dir_clr = val;
+        while (pbdrv_rproc_ev3_pru1_shared_ram.gpio_bank_01_dir_clr) {
+            // Wait for the PRU to process the command
```

Comment:
I suppose this is very fast, but just wanted to check if we can afford a blocking loop. Or maybe add a comment to indicate why it isn't an issue.

See also comment above. I suppose it is even less of an issue if we don't do this over and over. The color sensor will keep toggling a pin high and low hundreds of times per second, so call this loop as many times (not sure which bank these are on).